### DISCUSSION TITLE
30 second loop generating code

### DISCUSSION THREAD

**Discussion by arrowgent** [dyc3/steamguard-cli#165] [Show and tell]
to ease the process and timing of generating codes based on steam's generation @ :00 seconds and :30 seconds

here is a bash script with some rudimentary code to loop the process

enter your required variables at the top
then `chmod +x steamguard-cli.sh`
then `./steamguard-cli.sh`
it will loop every 30 seconds and copy the key to the clipboard

```
#!/bin/bash

#timer 30 seconds
#steamguard-cli uses SDA (windows) maFiles manifest (see .wine)


#your maFiles location
MAFILES="/home/$USER/.local/bin/steamguard-cli/maFiles"

#your passkey or decryption if required (should be required)
STEAMGUARD_CLI_PASSKEY=$(cat /A/TEXT/FILE/DIR/OR/OTHER/ENCRYPTION/METHOD/p.txt)

#obvious your steam username
USERNAME="YOUR USERNAME"

while true; do
# do code first time
exec /home/$USER/.local/bin/steamguard-cli/steamguard -m $MAFILES -u $USERNAME -p $STEAMGUARD_CLI_PASSKEY code | xclip -selection clipboard
echo "copied to clipboard";
xsel -b;

#first date in seconds
timer=$(date +%S)
#always true date = less than :60
    while [ $timer -le 60 ];
    do
# check date timer for :00 or :30 to start new code
            if [[ "echo $timer" =~ "00"|"30" ]];
            then
                exec /home/$USER/.local/bin/steamguard-cli/steamguard -m $MAFILES -u $USERNAME -p $STEAMGUARD_CLI_PASSKEY code | xclip -selection clipboard
                echo "generating new key";
                echo "copied to clipboard";
                xsel -b;
#new date seconds
timer=$(date +%S)
                echo $timer;
# timeout 30 seconds
                sleep 1;
            else
#new date seconds
timer=$(date +%S)
                echo $timer;
# seconds interval loop
                sleep 1;
            fi
    done
done
```

**Comment by arrowgent** [comment]
also noting it takes 5-7 seconds to generate a key using steamguard-cli

**Reply by dyc3** [reply]
You can use `--offline` to speed up code generation. By default, steamguard-cli reaches out to valve servers to get the time because the user's clock might be wrong.

**Reply by arrowgent** [reply]
awesome thanks. that will be faster than asking steam for the time. my time is always updated


checking with `--offline`  it still seems to take about 6 seconds
not sure why it needs this extra time

the SDA seems to happen instantly on `:00` and `:30` rollover, maybe it is pre-emptively generating the code


```57
58
59
00
INFO - reading manifest from /home/USER1/.local/bin/steamguard-cli/maFiles
INFO - Generating code for A USER
generating new key
copied to clipboard
XxXxX
06
07
```

**Reply by dyc3** [reply]
Using encryption will make it take longer to generate codes because steamguard-cli has to decrypt the appropriate file every time it is run, and the encryption is slow because of the PBKDF2 iterations. https://github.com/dyc3/steamguard-cli/blob/db6557ca0d0fbb18345f1be2b7c1d4efaa5f922f/src/encryption.rs#L66

**Reply by dyc3** [reply]
Also IIRC steam does give you some leeway with the codes it accepts. I think it'll accept the previous code and the current code in case the user is too slow typing it in.
